src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,623 @@
+                }
+                throw new UnsupportedOperationException();
+              }));
+    }
+  }
+
+  @Override
+  public BatchStat getBatchStatter() {
+    return new GrpcBatchStat();
+  }
+
+  @Override
+  public boolean canCreateSymlinkTree() {
+    return true;
+  }
+
+  @Override
+  public void createSymlinkTree(Map<PathFragment, PathFragment> symlinks, PathFragment symlinkTreeRoot)
+      throws ExecException, InterruptedException {
+    // The provided set of symbolic links may be too large to provide to
+    // the remote output service at once. Partition the symbolic links
+    // in groups of 1000, so that BatchCreateRequest messages remain
+    // small enough.
+    UnmodifiableIterator<List<Map.Entry<PathFragment, PathFragment>>> symlinkBatchIterator =
+        partition(symlinks.entrySet().iterator(), 1000);
+    boolean cleanPathPrefix = true;
+    while (symlinkBatchIterator.hasNext()) {
+      List<Map.Entry<PathFragment, PathFragment>> symlinksBatch = symlinkBatchIterator.next();
+      BatchCreateRequest.Builder builder = BatchCreateRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setPathPrefix(symlinkTreeRoot.relativeTo(currentRelativeOutputPath).toString());
+      // During the first iteration, we should ensure that any existing
+      // contents of the symlink tree directory are removed. This may be
+      // triggered by setting the 'clean_path_prefix' option.
+      builder.setCleanPathPrefix(cleanPathPrefix);
+      for (Map.Entry<PathFragment, PathFragment> symlink : symlinksBatch) {
+        PathFragment target = symlink.getValue();
+        if (target == null) {
+          // No target indicates an empty file needs to be created.
+          OutputFile.Builder fileBuilder = builder.addFilesBuilder();
+          fileBuilder.setPath(symlink.getKey().toString());
+          fileBuilder.setDigest(emptyDigest);
+          fileBuilder.setIsExecutable(true);
+        } else {
+          OutputSymlink.Builder symlinkBuilder = builder.addSymlinksBuilder();
+          symlinkBuilder.setPath(symlink.getKey().toString());
+          symlinkBuilder.setTarget(symlink.getValue().toString());
+        }
+      }
+      try {
+        channel.withChannelBlocking(
+            channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+                .batchCreate(builder.build()));
+      } catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+      cleanPathPrefix = false;
+    }
+  }
+
+  @Override
+  public void clean() throws ExecException, InterruptedException {
+    CleanRequest.Builder builder = CleanRequest.newBuilder();
+    builder.setOutputBaseId(outputBaseId);
+    try {
+      channel.withChannelBlocking(
+          channel -> RemoteOutputServiceGrpc.newBlockingStub(channel)
+              .clean(builder.build()));
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
```

Comment:
nit: That probably calls for some `ExecException`, does anything catch `UncheckedIOException` down the line?

Reply:
I have to confess I only added these UncheckedIOExceptions, because I wasn't sure what exception type ought to be returned here.

- It looks like createSymlinkTree() and clean() can already throw ExecException. Are you saying that ExecException is the right type to use here?
- startBuild() and sendFinalizeBuild() don't throw ExecException, but I guess that's also intended, because they don't run as part of executing a single action/target. Would BuildFailedException be the right exception type here?

Reply:
For `startBuild`, it throws `AbruptExitException` which is most likely what you want/what we generally use for such failures outside of execution. Probably the same would work for `sendFinalizeBuild`.